sap_hana/datadog_checks/sap_hana/sap_hana.py

```diff
@@ -0,0 +1,569 @@
+                        try:
+                            metric_info.append((name, float(value), column_type))
+                        except (ValueError, TypeError):  # no cov
+                            self.log.error('Non-numeric value `{}` for metric column `{}`'.format(value, name))
+                            break
+
+                # Only submit metrics if there were absolutely no errors - all or nothing.
+                else:
+                    for info in metric_info:
+                        metric, value, method = info
+                        getattr(self, method)(metric, value, tags=query_tags)
+
+    def iter_rows(self, query, implicit_values=True):
+        # https://github.com/SAP/PyHDB
+        with closing(self._conn.cursor()) as cursor:
+            cursor.execute(query.query)
+
+            # Re-use column access map for efficiency
+            result = {}
+
+            rows = cursor.fetchmany(self._row_limit)
+            while rows:
+                for row in rows:
+                    for column, value in zip(query.fields, row):
+                        # Undefined database sources may yield negative values.
+                        # This can also occur for metrics that have no value yet.
+                        if implicit_values and isinstance(value, (float, int)):
+                            value = positive(value)
+
+                        result[column] = value
+
+                    yield result
+
+                # Get next result set, if any
+                rows = cursor.fetchmany(self._row_limit)
+
+    def iter_rows_raw(self, query):
+        with closing(self._conn.cursor()) as cursor:
+            cursor.execute(query)
+
+            rows = cursor.fetchmany(self._row_limit)
+            while rows:
+                for row in rows:
+                    yield row
+
+                # Get next result set, if any
+                rows = cursor.fetchmany(self._row_limit)
+
+    def get_connection(self):
+        try:
+            connection = pyhdb.connection.Connection(
+                host=self._server, port=self._port, user=self._username, password=self._password, timeout=self._timeout
+            )
+            connection.connect()
+        except Exception as e:
+            error = str(e).replace(self._password, '*' * len(self._password))
+            self.log.error('Unable to connect to SAP HANA: {}'.format(error))
+            self.service_check(self.SERVICE_CHECK_CONNECT, self.CRITICAL, message=error, tags=self._tags)
+        else:
+            self.service_check(self.SERVICE_CHECK_CONNECT, self.OK, tags=self._tags)
+            return connection
+
+    def get_hana_hostname(self, hostname=None):
+        if self._use_hana_hostnames:
+            return hostname or self._master_hostname
+
+    @classmethod
+    def detect_used_views(cls):
+        try:
+            _, _, traceback = sys.exc_info()
+            query_class = traceback.tb_next.tb_next.tb_frame.f_locals['query'].__name__
```

Comment:
That's a bit too smart for its own good. Maybe you can iterate on the views themselves, and have a map of view to methods?